File: 118-PascalsTriangle/118-PascalsTriangle.js
// Last updated: 4/29/2026, 10:34:32 AM
/**
 * @param {number} numRows
 * @return {number[][]}
 */
// 2022-11-09
// 10:05
// O(n^2)
// O(n^2)
var generate = function(numRows) {
    const ans = [];
    const pascal = n => {
        if(n === 1) {
            return [1];
        } else if(n === 2) {
            return [1, 1];
        }
        const prev = pascal(n - 1);
        const arr = [1];
        for(let i = 0; i < prev.length - 1; i++) {
            arr.push(prev[i] + prev[i + 1]);
        }
        arr.push(1);
        return arr;
    }
    for(let i = 1; i <= numRows; i++) {
        ans.push(pascal(i));
    }
    return ans;
};

File: 118-PascalsTriangle/118-PascalsTriangle.py
# Last updated: 4/29/2026, 10:34:27 AM
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        def getValue(l, idx):
            if idx < 0 or idx >= len(l):
                return 0
            return l[idx]
        
        res = []
        def generateNextRow(prev_list):
            n = len(prev_list)
            arr = []
            for i in range(n+1):
                arr.append(getValue(prev_list, i-1) + getValue(prev_list, i))
            return arr
        row = None
        for i in range(1, numRows+1):
            if i == 1:
                row = [1]
            else:
                row = generateNextRow(row)
            res.append(row)
            # print(i)
            # print(row)
        return res